### PR TITLE
Avoid to use iDynTree vectors and matrices in FloatingBaseSystemDynamics

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
With this PR we reduce the memory allocation in `FloatingBaseSystemDynamics`. 
This PR is possible thanks to https://github.com/robotology/idyntree/pull/736

### CODE REVIEW

File: src/System/src/FloatingBaseSystemDynamics.cpp [hunk 1]
```diff
@@ -142,29 +141,17 @@ bool FloatingBaseDynamicalSystem::dynamics(const double& time,
...
-    if (!m_kinDyn->setRobotState(iDynTree::Transform(baseRot, basePos),
-                                 jointPos,
-                                 baseTwist,
-                                 jointVel,
-                                 m_gravity))
+    Eigen::Matrix4d baseTransform;
+    baseTransform.topLeftCorner<3,3>() = baseOrientation;
+    baseTransform.topRightCorner<3,1>() = basePosition;
+    baseTransform(3,3) = 1;
+    baseTransform.bottomLeftCorner<1,3>().setZero();
```
Comment (prashanthr05) [inline src/System/src/FloatingBaseSystemDynamics.cpp:148]: We could maybe add a utility method for this somewhere?

**Comment by traversaro** [comment]
Reading a bit all this could, perhaps for readability it could make sense to also define some functions ` iDynTree::make_span` (or some other name if less confusing, such as `iDynTree::asSpan`) in `iDynTree/Core/EigenHelpers.h` that take directly a single argument and return a span, so that all `iDynTree::make_span(jointPositions.data(), jointPositions.size())` can be simplified to `iDynTree::make_span(jointPositions)` or `iDynTree::asSpan(jointPositions)` ? For the span we could even make it just a template function that assume that the input type implement .data and size methods, to automatically cover also the std and YARP case. For the MatrixView probably we will need YARP and Eigen specific function to handle the row major and column major aspects.

**Comment by GiulioRomualdi** [comment]
> Reading a bit all this could, perhaps for readability it could make sense to also define some functions ` iDynTree::make_span` (or some other name if less confusing, such as `iDynTree::asSpan`) in `iDynTree/Core/EigenHelpers.h` that take directly a single argument and return a span, so that all `iDynTree::make_span(jointPositions.data(), jointPositions.size())` can be simplified to `iDynTree::make_span(jointPositions)` or `iDynTree::asSpan(jointPositions)` ? For the span we could even make it just a template function that assume that the input type implement .data and size methods, to automatically cover also the std and YARP case. For the MatrixView probably we will need YARP and Eigen specific function to handle the row major and column major aspects.

For the span it can be defined in `iDynTree/Core/Span.h` right? In the end `make_span` is a function that we defined

**Comment by traversaro** [comment]
> Noticed that that constructor is something that we defined. Indeed in the original implementation of span only these constructors are defined. So if you agree we can change it 

Ok for me!

File: src/System/src/FloatingBaseSystemDynamics.cpp (hunk 1 above)
Comment (GiulioRomualdi) [reply src/System/src/FloatingBaseSystemDynamics.cpp]: Done here: https://github.com/dic-iit/bipedal-locomotion-framework/pull/108/commits/27952e9fb75e543331dbcb6b2902da36579e07d3

**Comment by GiulioRomualdi** [comment]
@traversaro and @prashanthr05 let me know if you're fine with the latest modification, so I can merge the PR

**Review by traversaro — APPROVED** [review]

File: src/System/src/FloatingBaseSystemDynamics.cpp [hunk 2]
```diff
@@ -14,6 +14,18 @@
...
+                                       Eigen::Ref<const Eigen::Matrix3d> rotation)
+{
+    Eigen::Matrix4d transform;
+    transform.topLeftCorner<3,3>() = rotation;
+    transform.topRightCorner<3,1>() = position;
+    transform(3,3) = 1;
+    transform.bottomLeftCorner<1,3>().setZero();
+
+    return transform;
+}
```
Comment (prashanthr05) [inline src/System/src/FloatingBaseSystemDynamics.cpp:27]: We had added this function in `Conversions` https://github.com/dic-iit/bipedal-locomotion-framework/blob/43c8b35a2659af9dad910999a08c35766a95e473/src/Conversions/include/BipedalLocomotion/Conversions/CommonConversions.h#L25.
Would you not like to use it ?

File: src/System/src/FloatingBaseSystemDynamics.cpp (hunk 1 above)
Comment (prashanthr05) [reply src/System/src/FloatingBaseSystemDynamics.cpp]: Please see https://github.com/dic-iit/bipedal-locomotion-framework/pull/108#discussion_r527050397

File: src/System/src/FloatingBaseSystemDynamics.cpp (hunk 2 above)
Comment (GiulioRomualdi) [reply src/System/src/FloatingBaseSystemDynamics.cpp:27]: Because I forgot!

Comment (GiulioRomualdi) [reply src/System/src/FloatingBaseSystemDynamics.cpp]: https://github.com/dic-iit/bipedal-locomotion-framework/pull/108/commits/3673a8801b78e9f304d3133a5ab66b97d4bfa838

**Review by prashanthr05 — APPROVED** [review]

**Pull request merged**